src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -2794,23 +2834,33 @@ def _array_syntax_to_indexed(self, parent, loop_vars):
                                      found.
         '''
         assigns = parent.walk(Assignment)
-        # Check that the LHS of any assignment uses recognised array
-        # notation.
+        # Check that the LHS of any assignment that has Ranges on the RHS
+        # uses recognised array notation.
         for assign in assigns:
+            ranges = assign.rhs.walk(Range)
+            if not ranges:
+                # No array ranges on the RHS of this assignment
+                continue
             _ = self._array_notation_rank(assign.lhs)
```

Comment:
This resulted in a bug if a `where` contained e.g. `ll_mask = .true.` where `ll_mask` is an array. Since there is no array notation on either the LHS or RHS we assumed it was a scalar. I've changed the code so that we raise a NotImplementedError instead in this situation.

I've added a note about the duplicated functionality to #1576 and put a TODO in the code.